cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ImpersonationHandler.java

```diff
@@ -0,0 +1,105 @@
+import org.apache.twill.api.SecureStore;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+import org.jboss.netty.handler.codec.http.HttpRequest;
+import org.jboss.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedOutputStream;
+import java.io.DataOutputStream;
+import java.util.concurrent.Callable;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+
+/**
+ * Resolves the UGI for a given namespace, acquires the delegation tokens for that UGI,
+ * using {@link TokenSecureStoreUpdater}, and serializes these Credentials to a location.
+ *
+ * Response with the location to which the credentials were serialized to, as well as the UGI's short username
+ */
+// we don't share the same version as other handlers, so we can upgrade/iterate faster
+@Path("/v1")
+public class ImpersonationHandler extends AbstractHttpHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultUGIProvider.class);
+  private static final Gson GSON = new Gson();
+
+  private final DefaultUGIProvider ugiProvider;
+  private final TokenSecureStoreUpdater tokenSecureStoreUpdater;
+  private final LocationFactory locationFactory;
+
+  @Inject
+  public ImpersonationHandler(DefaultUGIProvider tokenRequester,
+                              TokenSecureStoreUpdater tokenSecureStoreUpdater,
+                              LocationFactory locationFactory) {
+    this.ugiProvider = tokenRequester;
+    this.tokenSecureStoreUpdater = tokenSecureStoreUpdater;
+    this.locationFactory = locationFactory;
+  }
+
+  @POST
+  @Path("/credentials")
+  public void getCredentials(HttpRequest request, HttpResponder responder) throws Exception {
+    String requestContent = request.getContent().toString(Charsets.UTF_8);
+    if (requestContent == null) {
+      throw new BadRequestException("Request body is empty.");
+    }
+    ImpersonationInfo impersonationInfo = GSON.fromJson(requestContent, ImpersonationInfo.class);
+
+    UserGroupInformation ugi = ugiProvider.getConfiguredUGI(impersonationInfo);
+    Credentials credentials = ImpersonationUtils.doAs(ugi, new Callable<Credentials>() {
+      @Override
+      public Credentials call() throws Exception {
+        SecureStore update = tokenSecureStoreUpdater.update(null, null);
+        return update.getStore();
+      }
+    });
+
+    // example: hdfs:///cdap/credentials
+    Location credentialsDir = locationFactory.create("credentials");
+    Preconditions.checkState(credentialsDir.mkdirs());
+
+    // the getTempFile() doesn't create the file within the directory that you call it on. It simply appends the path
+    // without a separator, which is why we manually append the "tmp"
+    // example: hdfs:///cdap/credentials/tmp.5960fe60-6fd8-4f3e-8e92-3fb6d4726006.credentials
+    Location credentialsFile = credentialsDir.append("tmp").getTempFile(".credentials");
```

Comment:
I think I raised the question about when this temp file will get removed and I didn't see it has been answer.